### PR TITLE
🎨 Add ability to filter learners by classroom #862

### LINKED ISSUE #862: 🎨 💄 Add ability to filter learners by classroom.
## User stories

As a user, I want to have the ability to sort users by class on the learners page.

---

Quick links: [Summary](#summary) | [Description](#description) | [Template](#template) | [Example](#example) | [Resources](#resources)

---

### Technical Description

Learners Page

- To achieve this you'll have to listen to the onchange event on the select option that triggers a filter method in your component. 
- The learner's data is managed by Mat-Table you can look into the documentation linked to understand how their filter works and integrate that into the filter method. : [filter-docs](https://material.angular.io/components/table/overview#filtering)

The learners can be found at `libs\features\convs-mgr\learners`.

### Acceptance Criteria

1. Changing the selection allows me to filter the learners by class.
2. Filter works with ungrouped users, marked as 'Default'.
3. All changes should be isolated in `libs\features\convs-mgr\learners`.

### PR DESCRIPTION
**Pull request by dennis-mburu**
# Description

This PR adds the ability to filter learners by class on the learners page.

Fixes #862 

## Type of change
- [x] New feature (non-breaking change which adds functionality)

# Screenshot (optional)

https://github.com/italanta/elewa/assets/65117575/9bbd8ba7-9a8b-4f1e-b04f-13073808946a

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings

### CODE REVIEW

**Review by LemmyMwaura — CHANGES_REQUESTED** [review]
Good work on this but you should customise it to work by updating the filterPredicate method that matTable comes with. This will allow us to add more complex filters

**Comment by dennis-mburu** [comment]
> Good work on this but you should customise it to work by updating the filterPredicate method that matTable comes with. This will allow us to add more complex filters

Changes Implemented. Kindly review @LemmyMwaura

File: libs/features/convs-mgr/learners/src/lib/pages/learners-page/learners-page.component.ts [hunk 1]
```diff
@@ -90,8 +90,26 @@ export class LearnersPageComponent implements OnInit, OnDestroy {
...
+  customFilterPredicate(learner: EnrolledEndUser, filter: string): boolean{
+    if (filter === "undefined" || filter === "null") {
+      return true; // No filter, include all rows
+    }
+
+    // Filter based on classId
+    const classCondition = !filter || learner.classId.toLowerCase().includes(filter);
+    
+    return classCondition
+  }
```
Comment (LemmyMwaura) [inline libs/features/convs-mgr/learners/src/lib/pages/learners-page/learners-page.component.ts:110]: When filter === 'Default", return users that have no class set. class = ""

File: libs/features/convs-mgr/learners/src/lib/pages/learners-page/learners-page.component.ts [hunk 2]
```diff
@@ -149,9 +168,10 @@ export class LearnersPageComponent implements OnInit, OnDestroy {
   }
 
   filterTable(event: Event, mode:string) {
+    const selectedValue = (event.target as HTMLSelectElement).value;
     switch (mode) {
       case 'class':
-        this.dataSource.filter = this.selectedClass.ClassName;
+        this.dataSource.filter = selectedValue.trim().toLowerCase() || ""
         break
     }
```
Comment (LemmyMwaura) [inline libs/features/convs-mgr/learners/src/lib/pages/learners-page/learners-page.component.ts:176]: Mode doesn't get passed here. Pass it in your custom filter predicate.

**Review by LemmyMwaura — CHANGES_REQUESTED** [review]

**Comment by dennis-mburu** [comment]
Both changes addressed. Kindly review @LemmyMwaura

**Review by LemmyMwaura — CHANGES_REQUESTED** [review]
Very good work on this. But the new update creates a lot more complexities than i would want on the search and the default search doesn't work. So let's just revert to our ealier implementation. I'll then find a way to customise it to use the filter predicate. 

For now make sure it work cleanly but with the initial implementation.

**Review by LemmyMwaura — APPROVED** [review]

**Pull request merged**